pojo/src/main/java/com/spotify/hamcrest/pojo/IsPojo.java

```diff
@@ -21,86 +21,116 @@
+          method.setAccessible(true);
+          @SuppressWarnings("unchecked")
+          final T returnValue = (T) method.invoke(self);
+          return returnValue;
+        },
+        returnValueMatcher);
   }
 
-  public IsPojo<A> where(String methodName, Matcher<?> returnValueMatcher) {
-    return whereWithoutTypeSafety(new MethodValueSupplier<>(methodName), returnValueMatcher);
+  public <T> IsPojo<A> where(
+      final MethodReference<A, T> methodReference,
+      final Matcher<T> returnValueMatcher) {
+    return where(
+        serializeLambda(methodReference).getImplMethodName(),
+        methodReference,
+        returnValueMatcher);
   }
 
-  public <T> IsPojo<A> where(ValueProvider<A, T> valueProvider, Matcher<T> valueMatcher) {
-    return whereWithoutTypeSafety(
-        new NamedLambdaValueSupplier<>(
-            serializeLambda(valueProvider).getImplMethodName(),
-            valueProvider),
-        valueMatcher);
+  private <T> IsPojo<A> where(
+      final String methodName,
+      final MethodReference<A, T> valueExtractor,
+      final Matcher<T> matcher) {
+    return toBuilder()
+        .methodHandler(methodName, MethodHandler.create(valueExtractor, matcher))
+        .build();
   }
 
   public IsPojo<A> withProperty(String property, Matcher<?> valueMatcher) {
     return where("get" + CaseFormat.LOWER_CAMEL.to(CaseFormat.UPPER_CAMEL, property),
-                 valueMatcher);
+        valueMatcher);
+  }
+
+  private static <A> Builder<A> builder(final Class<A> cls) {
+    return new AutoValue_IsPojo.Builder<A>().cls(cls);
   }
 
-  private IsPojo<A> whereWithoutTypeSafety(Function<A, ?> valueSupplier, Matcher<?> valueMatcher) {
-    return new IsPojo<>(cls,
-                        ImmutableMap.<Function<A, ?>, Matcher<?>>builder()
-                            .putAll(methodMatchers)
-                            .put(valueSupplier, valueMatcher)
-                            .build());
+  abstract Builder<A> toBuilder();
+
+  @AutoValue.Builder
+  abstract static class Builder<A> {
+
+    abstract Builder<A> cls(final Class<A> cls);
+
+    abstract ImmutableMap.Builder<String, MethodHandler<A, ?>> methodHandlersBuilder();
+
+    Builder<A> methodHandler(final String methodName, final MethodHandler<A, ?> handler) {
+      methodHandlersBuilder().put(methodName, handler);
+      return this;
+    }
+
+    abstract IsPojo<A> build();
   }
 
   @Override
   protected boolean matchesSafely(A item, Description mismatchDescription) {
-    if (!cls.isInstance(item)) {
-      mismatchDescription.appendText("not an instance of " + cls.getName());
+    if (!cls().isInstance(item)) {
+      mismatchDescription.appendText("not an instance of " + cls().getName());
       return false;
     }
 
-    final Map<Function<A, ?>, Consumer<Description>> mismatches = new LinkedHashMap<>();
+    final Map<String, Consumer<Description>> mismatches = new LinkedHashMap<>();
```

Comment:
If this is `String` now then the generics used in `DescriptionUtils` should be removed